src/SDKs/Azure.Core/data-plane/Azure.Core.Tests/ClientTestBase.cs

```diff
@@ -0,0 +1,64 @@
+﻿// Copyright (c) Microsoft Corporation. All rights reserved.
+// Licensed under the MIT License.
+
+using System;
+using System.Reflection;
+using Castle.DynamicProxy;
+using NUnit.Framework;
+
+namespace Azure.Core.Tests
+{
+    [TestFixture(true)]
+    [TestFixture(false)]
+    public class ClientTestBase
+    {
+        private static readonly ProxyGenerator _proxyGenerator = new ProxyGenerator();
+
+        private static readonly UseSyncMethodsInterceptor _interceptor = new UseSyncMethodsInterceptor();
+
+
+        public bool IsAsync { get; }
+
+        public ClientTestBase(bool isAsync)
+        {
+            IsAsync = isAsync;
+        }
+
+        public virtual TClient CreateClient<TClient>(params object[] args) where TClient: class
+        {
+
+            return WrapClient((TClient)Activator.CreateInstance(typeof(TClient), args));
+        }
+
+        public virtual TClient WrapClient<TClient>(TClient client) where TClient: class
+        {
+            if (ClientValidation<TClient>.Validated == false)
+            {
+                foreach (var methodInfo in typeof(TClient).GetMethods(BindingFlags.Instance | BindingFlags.Public))
+                {
+                    if (methodInfo.Name.EndsWith("Async") && !methodInfo.IsVirtual)
+                    {
+                        ClientValidation<TClient>.ValidationException = new InvalidOperationException($"Client type contains public non-virtual async method {methodInfo.Name}");
```

Comment:
I'm assuming this is a requirement to instrument?  I don't think we make any of our methods virtual by default.  Should this become part of the guidelines if so?

Also, why not `break` after we find the first problem method?

Reply:
> I'm assuming this is a requirement to instrument? I don't think we make any of our methods virtual by default. Should this become part of the guidelines if so?

It's part of mock-ability guidelines that are not added to .net guidelines yet.

> Also, why not break after we find the first problem method?

Good point